extensions/datastores/kudu/src/main/java/org/locationtech/geowave/datastore/kudu/operations/KuduRangeRead.java

```diff
@@ -1,189 +1,266 @@
+    }, "Kudu Query Executor").start();
+    return new CloseableIteratorWrapper<T>(new Closeable() {
+      @Override
+      public void close() throws IOException {
+        isCanceled.set(true);
       }
-      return new CloseableIteratorWrapper<>(() -> {
-      },
-          rowTransformer.apply(
-              rowMerging ? new GeoWaveRowMergingIterator(tmpIterator) : tmpIterator));
-    } else {
-      // Order the rows for data index query
-      final Map<ByteArray, GeoWaveRow> resultsMap = new HashMap<>();
-      while (concatIterator.hasNext()) {
-        RowResult r = concatIterator.next();
-        final byte[] d = r.getBinaryCopy(KuduDataIndexField.GW_PARTITION_ID_KEY.getFieldName());
-        resultsMap.put(
-            new ByteArray(d),
-            KuduDataIndexRow.deserializeDataIndexRow(r, visibilityEnabled));
+    }, new RowConsumer(results));
+  }
+
+  public void checkFinalize(
+      final AsyncKuduScanner scanner,
+      final Semaphore semaphore,
+      final BlockingQueue<Object> resultQueue,
+      final AtomicInteger queryCount) {
+    scanner.close();
+    semaphore.release();
+    if (queryCount.decrementAndGet() <= 0) {
+      try {
+        resultQueue.put(RowConsumer.POISON);
+      } catch (final InterruptedException e) {
+        LOGGER.error("Interrupted while finishing blocking queue, this may result in deadlock!");
       }
-      tmpIterator =
-          Arrays.stream(dataIds).map(d -> resultsMap.get(new ByteArray(d))).filter(
-              Objects::nonNull).iterator();
-      return new CloseableIteratorWrapper<>(() -> {
-      }, (Iterator<T>) tmpIterator);
     }
   }
 
+  public Deferred<Object> executeScanner(
+      final AsyncKuduScanner scanner,
+      final Semaphore semaphore,
+      final BlockingQueue<Object> resultQueue,
+      final AtomicInteger queryCount,
+      final AtomicBoolean isCanceled,
+      final boolean visibilityEnabled,
+      final Predicate<GeoWaveRow> filter,
+      final GeoWaveRowIteratorTransformer<T> rowTransformer,
+      final boolean rowMerging,
+      final short[] adapterIds) {
+    // Errback class
+    class QueryErrback implements Callback<Deferred<Object>, Exception> {
+      @Override
+      public Deferred<Object> call(final Exception e) {
+        LOGGER.warn("While scanning rows from kudu", e);
+        checkFinalize(scanner, semaphore, resultQueue, queryCount);
+        return Deferred.fromError(e);
+      }
+    }
+
+    // callback class
+    class QueryCallback implements Callback<Deferred<Object>, RowResultIterator> {
+      @Override
+      public Deferred<Object> call(RowResultIterator rs) {
+        Stream<GeoWaveRow> tmpStream;
+
+        if (rs == null || isCanceled.get()) {
+          checkFinalize(scanner, semaphore, resultQueue, queryCount);
+          return Deferred.fromResult(null);
+        }
+
+        if (rs.getNumRows() > 0) {
+          tmpStream = Streams.stream(rs.iterator()).map(r -> (GeoWaveRow) new KuduRow(r));
```

Comment:
Could you replace this anonymous function with `KuduRow::new`? Then, we can also remove the cast.